File: NetworkStubbingExperiment/NetworkStubbingExperimentUITests/NetworkStubbingExperimentUITests.m
#import <XCTest/XCTest.h>
#import "MockWebServer.h"
#import "EnvironmentKeys.h"

@interface NetworkStubbingExperimentUITests : XCTestCase
@property (nonatomic, strong) XCUIApplication *app;
@property (nonatomic, strong) MockWebServer *webServer;
@end

@implementation NetworkStubbingExperimentUITests

- (void)setUp {
    [super setUp];

    self.continueAfterFailure = NO;

    self.webServer = [[MockWebServer alloc] init];
    [self.webServer start];

    self.app = [[XCUIApplication alloc] init];
    self.app.launchEnvironment = @{ kServerURLEnvironmentKey: self.webServer.baseURL.absoluteString };
    [self.app launch];
}

- (void)tearDown {
    [super tearDown];
    [self.webServer stop];
}

- (void)testLoadsAndDisplaysTextFromServer {
    XCUIElement *loadedText = self.app.staticTexts[@"Hello World!"];
    [self expectationForPredicate:[NSPredicate predicateWithFormat:@"exists==YES"] evaluatedWithObject:loadedText handler:NULL];
    [self waitForExpectationsWithTimeout:5 handler:NULL];

    XCTAssert(loadedText.exists);
}

- (void)testTappingButtonSendsCheckinPost {
    NSString *expectedString = @"Checking in";
    [self keyValueObservingExpectationForObject:self.webServer keyPath:@"postedString" expectedValue:expectedString];

    [self.app.buttons[@"Check In"] tap];

    [self waitForExpectationsWithTimeout:5 handler:NULL];
    XCTAssertEqualObjects(self.webServer.postedString, expectedString);
}

@end

File: Swift/NetworkStubbingExperiment/NetworkStubbingExperimentUITests/NetworkStubbingExperimentUITests.swift
import XCTest

class NetworkStubbingExperimentUITests: XCTestCase {
    let app = XCUIApplication()
    let webServer = MockWebServer()

    override func setUp() {
        super.setUp()
        continueAfterFailure = false

        do {
            try webServer.start()
        } catch let error {
            XCTFail("Failed to start server: \(error)")
        }

        app.launchEnvironment = [ kServerURLEnvironmentKey: webServer.baseURL.absoluteString ]
        app.launch()
    }

    override func tearDown() {
        super.tearDown()
        webServer.stop()
    }

    func testLoadsAndDisplaysTextFromServer() {
        let loadedText = app.staticTexts["Hello World!"]
        let existsPredicate = NSPredicate(format: "exists==YES", argumentArray: nil)
        expectationForPredicate(existsPredicate, evaluatedWithObject: loadedText, handler: nil)
        waitForExpectationsWithTimeout(5, handler: nil)

        XCTAssert(loadedText.exists)
    }

    func testTappingButtonSendsCheckinPost() {
        let expectedString = "Checking in"
        keyValueObservingExpectationForObject(webServer, keyPath: "postedString", expectedValue: expectedString)

        app.buttons["Check In"].tap()

        waitForExpectationsWithTimeout(5, handler: nil)
        XCTAssertEqual(webServer.postedString, expectedString)
    }
}
